src/bun.js/node/types.zig

```diff
@@ -2331,34 +2331,104 @@ pub const Path = struct {
 
-        var stack_fallback_allocator = std.heap.stackFallback(
-            (32 * @sizeOf(string)),
-            heap_allocator,
-        );
-        var allocator = stack_fallback_allocator.get();
-        var out_buf: [bun.MAX_PATH_BYTES * 2]u8 = undefined;
-
-        var parts = allocator.alloc(string, args_len) catch unreachable;
-        defer allocator.free(parts);
-
         var arena = bun.ArenaAllocator.init(heap_allocator);
-        const arena_allocator = arena.allocator();
         defer arena.deinit();
 
-        var i: u16 = 0;
-        while (i < args_len) : (i += 1) {
-            parts[i] = args_ptr[i].toSlice(globalThis, arena_allocator).slice();
+        const cwd = Fs.FileSystem.instance.top_level_dir;
+        var out = cwd;
+        if (args_len > 0) {
+            const arena_allocator = arena.allocator();
+            var stack_fallback_allocator = std.heap.stackFallback(
+                (32 * @sizeOf(string)),
+                heap_allocator,
+            );
+            var allocator = stack_fallback_allocator.get();
+            const parts_len = args_len + 1;
+            var parts = allocator.alloc(string, parts_len) catch unreachable;
+            defer allocator.free(parts);
+            
+            parts[0] = cwd;
+            var i: u16 = 0;
+            while (i < args_len) : (i += 1) {
+                 parts[i + 1] = args_ptr[i].toSlice(globalThis, arena_allocator).slice();
+            }
+            if (isWindows) {
+                out = std.fs.path.resolveWindows(arena_allocator, parts) catch "";
+            } else {
+                out = std.fs.path.resolve(arena_allocator, parts) catch "";
+            }
         }
+        var outStr = JSC.ZigString.init(out);
+        if (arena.state.buffer_list.first != null)
+            outStr.setOutputEncoding();
 
-        var out: JSC.ZigString = if (!isWindows)
-            JSC.ZigString.init(strings.withoutTrailingSlash(PathHandler.joinAbsStringBuf(Fs.FileSystem.instance.top_level_dir, &out_buf, parts, .posix)))
-        else
-            JSC.ZigString.init(strings.withoutTrailingSlashWindowsPath(PathHandler.joinAbsStringBuf(Fs.FileSystem.instance.top_level_dir, &out_buf, parts, .windows)));
+        return outStr.toValueGC(globalThis);
+    }
 
-        if (arena.state.buffer_list.first != null)
-            out.setOutputEncoding();
+    fn toNamespacedPathPosix(path: []const u8) []const u8 {
+        return path;
+    }
+
+    // Based on Node's path.win32.toNamespacedPath:
+    // https://github.com/nodejs/node/blob/6ae20aa63de78294b18d5015481485b7cd8fbb60/lib/path.js#L622
+    fn toNamespacedPathWindows(path: []const u8) []const u8 {
+        var arena = bun.ArenaAllocator.init(heap_allocator);
+        const arena_allocator = arena.allocator();
+        defer arena.deinit();
 
-        return out.toValueGC(globalThis);
+        const resolvedPath = std.fs.path.resolveWindows(arena_allocator, &.{ path }) catch "";
```

Comment:
we shouldn't use the standard library function here

- it may not work 1:1 with how node `path.resolve` works (which has a ton of stupid quirks, especially on windows)
- this part of the operation really doesn't need to allocate. for small / temporary sizes it is usually better to stack allocate these.

you can use a temporary path buffer and the `PathHandler.joinAbsStringBuf` api

---

also as a further optimization to reduce the number of copies done:

this operation ever takes one of three paths:
- return the string as-is
- return the string with the prefix '\\?\UNC\'
- return the string with the prefix '\\?\'

we can have one buffer and write the resolved path 8 bytes into the start of it

```zig
var buf: bun.PathBuffer = undefined; // PathBuffer is alias for [MAX_PATH_BYTES]u8;
// actually, this should be a parameter, so that returning a slice from it will not be invalid

const unc_prefix = "\\\\?\\UNC\\";
const device_prefix = "\\\\?\\";

// write the path 
const resolved = strings.withoutTrailingSlashWindowsPath(
  PathHandler.joinAbsStringBuf(
    Fs.FileSystem.instance.top_level_dir, // cwd
    buf[unc_prefix.len..],
    &.{ input },
    .windows
  )
);
// joinAbsStringBuf may return a pointer to the `cwd` untouched
// otherwise, this lets us skip copying the path an extra time
if (resolved.ptr != buf[unc_prefix.len..].ptr) {
  @memcpy(buf[unc_prefix.len..][0..resolved.len], resolved);
}

// do logic to find what prefix we need, then write the prefix before
// byte number `unc_prefix.len`, and then return the slice starting there
```

Reply:
Updated implementations.